cmd/ebrelayer/relayer/prophecyHandler.go

```diff
@@ -0,0 +1,193 @@
+	"sync"
+	"syscall"
+	"time"
+
+	cosmosbridge "github.com/Sifchain/sifnode/cmd/ebrelayer/contract/generated/bindings/cosmosbridge"
+	"github.com/Sifchain/sifnode/cmd/ebrelayer/internal/symbol_translator"
+	"github.com/Sifchain/sifnode/cmd/ebrelayer/txs"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/ethereum/go-ethereum/ethclient"
+	tmClient "github.com/tendermint/tendermint/rpc/client/http"
+	"google.golang.org/grpc"
+
+	ethbridgetypes "github.com/Sifchain/sifnode/x/ethbridge/types"
+	oracletypes "github.com/Sifchain/sifnode/x/oracle/types"
+)
+
+const wakeupTimer = 60
+
+// StartProphecyHandler start Cosmos chain subscription and process prophecy completed message
+func (sub CosmosSub) StartProphecyHandler(txFactory tx.Factory, completionEvent *sync.WaitGroup, symbolTranslator *symbol_translator.SymbolTranslator) {
+	defer completionEvent.Done()
+	time.Sleep(time.Second)
+	client, err := tmClient.New(sub.TmProvider, "/websocket")
+	if err != nil {
+		sub.SugaredLogger.Errorw("failed to initialize a sifchain client.",
+			errorMessageKey, err.Error())
+		completionEvent.Add(1)
+		go sub.Start(txFactory, completionEvent, symbolTranslator)
+		return
+	}
+
+	if err := client.Start(); err != nil {
+		sub.SugaredLogger.Errorw("failed to start a sifchain client.",
+			errorMessageKey, err.Error())
+		completionEvent.Add(1)
+		go sub.Start(txFactory, completionEvent, symbolTranslator)
+		return
+	}
+
+	defer client.Stop() //nolint:errcheck
+
+	t := time.NewTicker(time.Second * wakeupTimer)
+
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
+	defer close(quit)
+
+	for {
+		select {
+		case <-quit:
+			sub.SugaredLogger.Warn("we receive the quit signal and exit")
+			return
+
+		case <-t.C:
+			sub.SugaredLogger.Info("timer triggered, start to check cosmos message")
+			sub.handleNewProphecyCompleted(client)
+		}
+	}
+}
+
+// Get all not processed Prophecy via rpc and handle them in batch
+func (sub CosmosSub) handleNewProphecyCompleted(client *tmClient.HTTP) {
+	// Start Ethereum client
+	ethClient, err := ethclient.Dial(sub.EthProvider)
+	if err != nil {
+		sub.SugaredLogger.Errorw("failed to connect to Ethereum node.",
+			errorMessageKey, err.Error())
+		return
+	}
+
+	cosmosBridgeAddress, err := txs.GetAddressFromBridgeRegistry(ethClient, sub.RegistryContractAddress, txs.CosmosBridge, sub.SugaredLogger)
+	if err != nil {
+		sub.SugaredLogger.Errorw("failed to get the cosmos bridge address.",
+			errorMessageKey, err.Error())
+		return
+	}
+
+	// get the last global nonce from smart contract
+	lastSubmittedNonce, err := GetLastNonceSubmitted(ethClient, cosmosBridgeAddress, sub.SugaredLogger)
+	if err != nil {
+		sub.SugaredLogger.Errorw("failed to get the last submitted nonce.",
+			errorMessageKey, err.Error())
+		return
+	}
+
+	prophecyInfoArray := GetAllProphciesCompleted(sub.TmProvider, sub.NetworkDescriptor, lastSubmittedNonce.Uint64()+1)
+
+	batches := (len(prophecyInfoArray) + 4) / 5
```

Comment:
We need to be clearer on why these numbers are needed. We should use a modulo operator here instead of just add 4.

Is this code supposed to round up @juniuszhou?